src/us/kbase/auth2/service/ui/Login.java

```diff
@@ -171,36 +174,40 @@ public Response loginStart(
 						PROVIDER_RETURN_EXPIRATION_SEC))
 				// will remove redirect cookie if redirect isn't set and one exists
 				.cookie(getRedirectCookie(redirect, PROVIDER_RETURN_EXPIRATION_SEC))
+				.cookie(getEnvironmentCookie(environment, UIPaths.LOGIN_ROOT,
+						PROVIDER_RETURN_EXPIRATION_SEC))
 				.build();
 	}
 	
-	private URL getRedirectURL(final String redirect)
-			throws AuthStorageException, IllegalParameterException {
+	private URL getRedirectURL(final String environment, final String redirect)
+			throws AuthStorageException, IllegalParameterException, NoSuchEnvironmentException {
 		final URL retRedirect;
 		if (nullOrEmpty(redirect)) {
 			retRedirect = null;
```

Comment:
At the time, I was trying to follow https://www.tomdalling.com/blog/coding-tips/coding-tip-have-a-single-exit-point/

I've since decided it's a silly rule (although returns in the middle of a long method or in a big nested blow are a bad idea) and I've changed it.